### ISSUE TITLE
Missing events of questions in a group shown on a field-list with relevance evaluation in the audit log

### ISSUE THREAD

**Issue by dbemke** [getodk/collect#5564]
#### ODK Collect version
the store version 2023.1.2, the master version 5dfd952592886ac948a694e926e2ec14084b3ae0
#### Android version
10
#### Device used
Redmi 9T
#### Problem description
The events connected with questions in a group shown on a field-list with relevance evaluation are missing in the audit log.

![relevantaditlog](https://user-images.githubusercontent.com/56479916/235148686-e5d620b0-e334-49bc-9045-436bb0cf50e8.png)

#### Steps to reproduce the problem
1. Go to Audit relevant constraint fieldlist form.
[Audit relevant constraint fieldlist.xlsx.txt](https://github.com/getodk/collect/files/11353461/Audit.relevant.constraint.fieldlist.xlsx.txt)
2. Fill the form.
3. Check the csv file with the audit log if there is an event for "relevanttobunch” event in the first group.
#### Expected behavior
All events should be present in the audit log.

**Comment by seadowg** [comment]
@dbemke could you provide an example of what the audit log should look like? I'm don't understand the problem here.

**Comment by dbemke** [comment]
Sorry I could have added more detailed steps.

**Steps to reproduce:**
1. Go to Audit relevant constraint fieldlist form.
2. In the first question enter 3.
3. "**Relevant bunch number” question appears, fill answer ( I entered "aaaaaa”).**
4. Tap next.
5. Fill other questions ( I entered 4, 5, 6, 7 in the following questions).
6. Save the form.

In the audit log  there isn't a question event between "3" (answer) and "4". So the question that appears in step 3 is missing.
![5564](https://github.com/getodk/collect/assets/56479916/1d20927b-8611-4817-86c4-3e3daaecd3ec)

**Comment by seadowg** [comment]
Ah I see! Could you try the same form/steps but remove the relevance from the question and see if you get a different log?

**Comment by dbemke** [comment]
You mean form with 2 questions on a field-list (without any constraint/required/relevance)?

**Comment by seadowg** [comment]
> You mean form with 2 questions on a field-list (without any constraint/required/relevance)?

Yup!

**Comment by dbemke** [comment]
[Audit field-list .xlsx.txt](https://github.com/getodk/collect/files/12302266/Audit.field-list.xlsx.txt)
q1 and q2 (questions on a field-list are missing) only the group event is visible in the audit log. Wasn't there a separate issue for questions on a field-list?
![auditFieldList](https://github.com/getodk/collect/assets/56479916/a2682198-90e2-4704-bd17-17fb0e892828)

**Comment by grzesiek2010** [comment]
We need to decide what we really want in this case.
If we have a group of questions on one page we log them first and then finalize (adding end time, answers for example) when navigating to another question. If the there are questions that appear/disappear depending on other questions then those events are not logged.

This is going to be tricky but we should try to add/remove question events when new questions appear/disappear.
Do you agree @lognaturel or maybe you think we should solve it in a different way?

**Comment by lognaturel** [comment]
@grzesiek2010 thanks for that explanation, it makes sense.

Am I understanding correctly that currently we always include events for questions in a field list as long as they are visible at the time that the field list is entered? That's always the case, even  when change tracking is not enabled? I ask because it might be better and more expected not to include any questions in a field list when change tracking is not enabled. Those events don't add any information and they make analysis more difficult -- if you want to do something like compute total time spent in the form you have to be careful not to count the time spent in the field list multiple times.

In the case of change tracking, we definitely do need question events in order to be able to capture the value change. Again, if we use the same start and end times as the enclosing group, that makes analysis harder.

An alternative to what you describe, @grzesiek2010, would be to forget about the group event and instead track question events within a field list based on focus. That would mean doing things like considering a question event "done" when the user interacts with another question. There could be multiple question events for the same question within a field list being shown. I don't have a good sense of how much work that would be but it does feel like it would be more useful.

**Comment by grzesiek2010** [comment]
> Am I understanding correctly that currently we always include events for questions in a field list as long as they are visible at the time that the field list is entered? That's always the case, even when change tracking is not enabled?

No, we do that only if change tracking is enabled so we don't have to change anything here. We might need to add a test for that cases because it looks like we don't have any.

> An alternative to what you describe, @grzesiek2010, would be to forget about the group event and instead track question events within a field list based on focus. That would mean doing things like considering a question event "done" when the user interacts with another question. There could be multiple question events for the same question within a field list being shown. I don't have a good sense of how much work that would be but it does feel like it would be more useful.

It sounds difficult... I was rather thinking about utilizing [odkView.removeWidgetAt(i)](https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/FormFillingActivity.java#L2539) and [odkView.addWidgetForQuestion(questionsAfterSave[i], i)](https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/FormFillingActivity.java#L2548C17-L2548C71)

**Comment by lognaturel** [comment]
> It sounds difficult

Likely true. What about only including question events for questions in field lists that have had value changes AND omitting start and end time for those? The times would be redundant with the group's times anyway.

**Comment by grzesiek2010** [comment]
> What about only including question events for questions in field lists that have had value changes AND omitting start and end time for those? The times would be redundant with the group's times anyway.

It already works like that (apart from omitting start and end time):

```
    /*
    Question which is in field-list group should be logged only if tracking changes option is
    enabled and its answer has changed
    */
    private boolean shouldEventBeLogged(AuditEvent aev) {
        if (aev.getAuditEventType().equals(AuditEvent.AuditEventType.QUESTION) && formController != null) {
            return !formController.indexIsInFieldList(aev.getFormIndex())
                    || (aev.hasNewAnswer() && auditConfig.isTrackingChangesEnabled());
        }
        return true;
    }
```

The problem is that we log question events for a group immediately once that group is displayed. Then when a user navigates away from that screen the questions events are completed, and filtered and only those are kept that have new values.
However if in that group you have questions that appear/disappear based on answers from other questions they won't be added/removed because as I said we log the questions events immediately once that group is displayed and do not update anything when `updateFieldListQuestions` method is called. 

So that's why I said above that probably we should:
> I was rather thinking about utilizing [odkView.removeWidgetAt(i)](https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/FormFillingActivity.java#L2539) and [odkView.addWidgetForQuestion(questionsAfterSave[i], i)](https://github.com/getodk/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/FormFillingActivity.java#L2548C17-L2548C71)
